#R
#GJR
#4/14/2021
# Test whether transcripts DE across stages (or with stageXline interactions) AND transcripts with stageXtreatment interactions tend to exhibit higher tissue specificity than expected by chance



#index for transcripts with stage effects or stageXline effects, but not stageXtreatment effects
ind<- (stage$FDR < 0.05 | stageByLine$FDR < 0.05 )  & stageByTreat$FDR > 0.05  

#unique ids from above
DEids<-unique(as.character(stage$genes[ind]))

#tissue specificity per gene estimated from flyAtlas 2.0 data
#Generated in 'calculateTauForTissueSpecificity.R'
tau<-read.table('/media/raglandlab/ExtraDrive1/dmel/FlyAtlas2_data/tauEstimatesPerFBgn.txt',header=T,stringsAsFactors = F,row.names=NULL)

#####################################################################
#create new data frame with updated flybase ids mapped to tau values#
#####################################################################
#create list of all ids, then need to submit to flybase id converter manually 
#allIds<-unique(c(DEids,tau$FBgn))
#write.table(allIds,'allIds.txt',sep="\t",quote=F,row.names=F)

#conversion table from flybase FB2021_02, 4/14/2021
#generated by submitting ids to: https://flybase.org/convert/id
idCon<-read.table('/media/raglandlab/ExtraDrive1/dmel/RNAseqStudy/PhilRnaSeqAnalysis/allIdsConverted.txt',header=T,sep="\t",row.names=NULL,stringsAsFactors=F,quote="\"",comment.char="")
names(idCon)[1]<-'FBgn'
idCon<-idCon[!duplicated(idCon$FBgn),]
tau<-merge(tau,idCon)

#####################################################################
## analyze stage and stageXline interaction transcripts #############
#####################################################################


DE<-data.frame(FBgn=DEids)
DE<-merge(DE,idCon)

ind<-tau$currentId %in% DE$currentId
pointEstTau<-median(tau$tau[ind],na.rm=T) #0.8802269

nGenes=sum(ind) #10931
nPerms<-10000
permutedTaus<-vector(length=nPerms)
for (i in 1:nPerms) {
  permutedTaus[i]<-median(sample(tau$tau,nGenes),na.rm=T)
}  
ecdf(permutedTaus)(pointEstTau) # 0 in 10000 perms with median taus <= ~0.88 p < 0.0001
mean(permutedTaus) # 0.9031893
#So, median tau for transcripts with stage effects or stageXline effects, but not stageXtreatment effects, is lower (less stage specific) than the expectation
#This may be a power issue, where tissue specific genes

#####################################################################
## analyze stageXtreatment interaction transcripts ##################
#####################################################################

ind<- stageByTreat$FDR < 0.05

 

DEids<-unique(as.character(stageByTreat$genes[ind]))
DE<-data.frame(FBgn=DEids)
DE<-merge(DE,idCon)

ind<-tau$currentId %in% DE$currentId
pointEstTau<-median(tau$tau[ind],na.rm=T) #0.82

nGenes=sum(ind) #849
nPerms<-10000
permutedTaus<-vector(length=nPerms)
for (i in 1:nPerms) {
  permutedTaus[i]<-median(sample(tau$tau,nGenes),na.rm=T)
}  
ecdf(permutedTaus)(pointEstTau) # 0 in 10000 perms with median taus <= ~0.82 p < 0.0001
mean(permutedTaus) #0.9028003
#same result as above, median tau for transcripts with stageXtreatment effects is lower than the expectation

############################################################################################################
## by contrast, transcripts with treatment effects but no interaction have roughly expected tau values #####
############################################################################################################

ind<-treatment$FDR < 0.05 & stageByTreat$FDR > 0.05
DEids<-unique(as.character(stageByTreat$genes[ind]))
DE<-data.frame(FBgn=DEids)
DE<-merge(DE,idCon)

ind<-tau$currentId %in% DE$currentId
pointEstTau<-median(tau$tau[ind],na.rm=T) #0.9020575

nGenes=sum(ind)
nPerms<-10000
permutedTaus<-vector(length=nPerms)
for (i in 1:nPerms) {
  permutedTaus[i]<-median(sample(tau$tau,nGenes),na.rm=T)
}  
ecdf(permutedTaus)(pointEstTau) #  p = 0.4909
mean(permutedTaus) #0.9028003